### PR TITLE
Kickstart the usage of `let_chains`

### PR DESCRIPTION
**Pull request by c410-f3r**
The time has come

### CODE REVIEW

**Review by flip1995 — COMMENTED** [review]
I just though about doing this today :smile: 

This is great. With more of this, we should be able to get rid of `if_chain!` completely, right?

File: clippy_lints/src/blocks_in_if_conditions.rs [hunk 1]
```diff
@@ -57,24 +56,23 @@ impl<'a, 'tcx> Visitor<'tcx> for ExVisitor<'a, 'tcx> {
...
-                if let ExprKind::MethodCall(_, [self_arg, ..], _) = &parent.kind;
-                let caller = self.cx.typeck_results().expr_ty(self_arg);
-                if let Some(iter_id) = self.cx.tcx.get_diagnostic_item(sym::Iterator);
-                if implements_trait(self.cx, caller, iter_id, &[]);
-                then {
-                    return;
-                }
+            if let Some(parent) = get_parent_expr(self.cx, expr)
+                && let ExprKind::MethodCall(_, [self_arg, ..], _) = &parent.kind
+                &&let caller = self.cx.typeck_results().expr_ty(self_arg)
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply clippy_lints/src/blocks_in_if_conditions.rs:61]: Not yet -> https://github.com/rust-lang/rustfmt/issues/5177

I guess it will be tackled eventually

**Comment by flip1995** [comment]
> also serves to battle-test the feature

There's probably no better place to test this than Clippy. It's ``if``s all the way down 😄

File: clippy_lints/src/attrs.rs [hunk 2]
```diff
@@ -264,23 +263,19 @@ declare_lint_pass!(Attributes => [
...
-                for item in items {
-                    if_chain! {
-                        if let NestedMetaItem::MetaItem(mi) = &item;
-                        if let MetaItemKind::NameValue(lit) = &mi.kind;
-                        if mi.has_name(sym::since);
-                        then {
-                            check_semver(cx, item.span(), lit);
-                        }
-                    }
+        if let Some(items) = &attr.meta_item_list() && let Some(ident) = attr.ident() {
```
Comment (camsteffen) [inline clippy_lints/src/attrs.rs:266]: I think `&& let` should always wrap to the next line. Does rustfmt allow either way?

**Comment by camsteffen** [comment]
I'm gonna miss our dear friend `if_chain!`. But this is really awesome! Congrats @c410-f3r for landing this feature in rustc!

**Comment by c410-f3r** [comment]
Closing due to formatting concerns. If desired, feel free to use any content of this PR for posterior submissions

**Comment by camsteffen** [comment]
I guess it makes sense to wait for rustfmt support? We already don't have formatting in `if_chain!`, although the lack of support is less surprising in a macro. Also formating by hand feels somewhat harder to get "right" with let-chains.

**Comment by xFrednet** [comment]
I think we can also accept PRs beforehand. As you said, `if_chain!`s also don't support formatting. In this case, we will even get formatting in the future in the worst case there will be one commit to align the formatting with rustfmt and that's it. :upside_down_face:

**Comment by camsteffen** [comment]
> I also think we can accept PRs beforehand.

I was mainly giving reasons to not migrate yet. But I don't have a strong opinion.

**Comment by xFrednet** [comment]
Sorry, in that case I misunderstood you :sweat_smile:. I thought it would be good to merge one PR like this one, to test the feature and to make sure that it work for us. Formatting is less important for a test. We can see if this causes any regressions in Nightly and there would still be time to fix it. However, I also don't have a strong opinion on this either xD

**Comment by flip1995** [comment]
I guess we could merge a PR like this. But it doesn't give any benefit _**for Clippy**_ until it can be automatically formatted. If we merge a PR like this, I would first limit it to `clippy_utils` and leave `clippy_lints` as-is for now.

**Comment by camsteffen** [comment]
In any case, I got the ball rolling for rustfmt support! rust-lang/rustfmt#5203

**Pull request closed without merging**